### PR TITLE
Concept id hints

### LINKED ISSUE #163: Add OMOP concept_id hints in RiaH
For RiaH it would be useful to have hints for concepts, e.g. gender_concept_id can only have two values; 8507 (MALE) and 8531 (FEMALE). Similar for all the type_concept_ids. During the mapping process one would always need to look these up, so it would save effort if the user is presented with the options in the RiaH interface.

@clairblacketer Do you have a list of fields that are associated with certain (domain of) concepts?

### PR DESCRIPTION
**Pull request by MaximMoinat**
Display concept id hints when selecting target fields, e.g. the available gender concepts for the `gender_concept_id` and the type concepts. The concept_id hints are loaded from a separate csv file, which has to kept up to date with vocabulary releases.

This PR introduces a `TargetFieldPanel` , to display the CDM field details including concept_id hints. This formats the values table specifically for concept id hints.
https://github.com/thehyve/OHDSI-WhiteRabbit/blob/9e3673dbe8bc2a14def8ef17104d39d1e4f6455d/rabbitinahat/src/main/java/org/ohdsi/rabbitInAHat/DetailsPanel.java#L479-L499

Closes #163

### CODE REVIEW

File: rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+                this.put(tableName, fieldName, concept);
+            }
+        } catch (IOException e) {
+            throw new IOException("Could not load concept_id hints: " + e.getMessage());
+        }
+    }
+
+    public void put(String targetTable, String targetField, Concept concept) {
+        if (!this.conceptMap.containsKey(targetTable)) {
```
Comment (blootsvoets) [inline rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java]: This sequence can be replaced with
```java
table = conceptMap.computeIfAbsent(targetTable, t -> new HashMap<>());
field = table.computeIfAbsent(targetField, f -> new ArrayList<>());
field.add(concept)
```

File: rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public List<Concept> get(String targetTable, String targetField) {
+        if (!this.containsKey(targetTable, targetField)) {
+            return null;
+        }
+        return this.conceptMap.get(targetTable).get(targetField);
+    }
+
+    private boolean containsKey(String targetTable, String targetField) {
```
Comment (blootsvoets) [inline rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java]: Likewise here.

File: rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            this.conceptMap.put(targetTable, new HashMap<>());
+        }
+        if (!this.conceptMap.get(targetTable).containsKey(targetField)) {
+            this.conceptMap.get(targetTable).put(targetField, new ArrayList<>());
+        }
+        this.conceptMap.get(targetTable).get(targetField).add(concept);
+    }
+
+    public List<Concept> get(String targetTable, String targetField) {
+        if (!this.containsKey(targetTable, targetField)) {
```
Comment (blootsvoets) [inline rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java]: possible alternative implementation of method:
```java
return conceptMap.getOrDefault(targetTable, Collections.emptyMap()).get(targetField);
```

File: rabbitinahat/src/main/java/org/ohdsi/rabbitInAHat/DetailsPanel.java [hunk 4]
```diff
@@ -315,14 +327,25 @@ public void removeUpdate(DocumentEvent e) {
 	private class FieldPanel extends JPanel implements DocumentListener {
 
 		private static final long	serialVersionUID	= -4393026616049677944L;
-		private JLabel				nameLabel			= new JLabel("");
-		private JLabel				rowCountLabel		= new JLabel("");
-		private DescriptionTextArea description			= new DescriptionTextArea ("");
-		private SimpleTableModel	valueTable			= new SimpleTableModel("Value", "Frequency", "Fraction");
-		private JTextArea			commentsArea		= new JTextArea();
-		private Field				field;
+		JLabel				nameLabel;
```
Comment (blootsvoets) [inline rabbitinahat/src/main/java/org/ohdsi/rabbitInAHat/DetailsPanel.java]: The outlining of this does not seem correct. Within a sentence, please use spaces to do outlining, or forego outlining altogether.

File: rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        this.conceptMap = new HashMap<>();
+    }
+
+    ConceptsMap(String filename) throws IOException{
+        this();
+        this.load(filename);
+    }
+
+    private void load(String filename) throws IOException{
+        InputStream conceptStream = Database.class.getResourceAsStream(filename);
```
Comment (blootsvoets) [inline rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java]: You can use a try-with-resources block here to ensure that the stream is closed again. You can add multiple statements to the block, e.g. `try (InputStream a = ...; InputStreamReader r= new InputStreamReader(a)) {`

File: rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java (hunk 1 above)
Comment (MaximMoinat) [reply rabbit-core/src/main/java/org/ohdsi/rabbitInAHat/dataModel/ConceptsMap.java]: That is really elegant!

**Pull request merged**